README.md
# acme_certificate 1.1.1

**THIS ROLE IS DEPRECATED! PLEASE USE THE [COLLECTION VERSION felixfontein.acme](https://galaxy.ansible.com/felixfontein/acme) (https://github.com/felixfontein/ansible-acme/) INSTEAD!**

Allows to obtain certificates from Let's Encrypt with minimal interaction with the webserver. Most code is executed on the controller, and the account key is never send to the nodes.

The role can be installed via [Ansible Galaxy](https://galaxy.ansible.com/felixfontein/acme_certificate):

    ansible-galaxy install felixfontein.acme_certificate

For changes in this role, see [the changelog](CHANGELOG.md).

## Description

This is an [Ansible](https://github.com/ansible/ansible) role which can use any CA supporting the ACME protocol, such as [Let's Encrypt](https://letsencrypt.org/) or [Buypass](https://www.buypass.com/ssl/products/acme), to issue TLS/SSL certificates for your server. This role requires Ansible 2.8.3 or newer and is based on the [acme_certificate module](https://docs.ansible.com/ansible/latest/acme_certificate_module.html) coming with Ansible.

The main advantage of this approach over others is that *almost no code is executed on your webserver*: only when you use HTTP challenges, files need to be copied onto your webserver, and afterwards deleted from it. Everything else is executed on your local machine!

(This does not cover installing the certificates, you have to do that yourself in another role.)

## Requirements

Requires the Python [cryptography](https://pypi.org/project/cryptography/) library installed on the controller, available to the Python version used to execute the playbook. If `cryptography` is not installed, a recent enough version of [PyOpenSSL](https://pypi.org/project/pyOpenSSL/) is currently supported as a fallback by the Ansible `openssl_privatekey` and `openssl_csr` modules.

The `openssl` binary must also be available in the executable path on the controller. It is needed by the `acme_certificate` module in case `cryptography` is not installed, and it is used for certificate chain validation.

If DNS challenges are used, there can be other requirements depending on the DNS provider. For example, for Amazon's Route 53, the Ansible `route53` module requires the Python `boto` package.

## Account Key Setup

You can create an account key using the `openssl` binary as follows:

    # RSA 4096 bit key
    openssl genrsa 4096 -out keys/acme-account.key
    # ECC 256 bit key (P-256)
    openssl ecparam -name prime256v1 -genkey -out keys/acme-account.key
    # ECC 384 bit key (P-384)
    openssl ecparam -name secp384r1 -genkey -out keys/acme-account.key

With Ansible, you can use the `openssl_privatekey` module as follows:

    - name: Generate RSA 4096 key
      openssl_privatekey:
        path: keys/acme-account.key
        type: RSA
        size: 4096
    - name: Generate ECC 256 bit key (P-256)
      openssl_privatekey:
        path: keys/acme-account.key
        type: ECC
        curve: secp256r1
    - name: Generate ECC 384 bit key (P-384)
      openssl_privatekey:
        path: keys/acme-account.key
        type: ECC
        curve: secp384r1

Make sure you store the account key safely. As opposed to certificate private keys, there is no need to regenerate it frequently, and it makes recovation of certificates issued with it very simple.

## Role Variables

Please note that from May 2020 on, all variables must be prefixed with `acme_certificate_`. For some time, the module will still use the old (short) variable names if the longer ones are not defined. Please upgrade your role usage as soon as possible.

These are the main variables:

- `acme_certificate_acme_account`: Path to the private ACME account key. Must always be specified.
- `acme_certificate_acme_email`: Your email address which shall be associated to the ACME account. Must always be specified.
- `acme_certificate_algorithm`: The algorithm used for creating private keys. The default is `"rsa"`; other choices are `"p-256"`, `"p-384"` or `"p-521"` for the NIST elliptic curves `prime256v1`, `secp384r1` and `secp521r1`, respectively.
- `acme_certificate_key_length`: The bitlength to use for RSA private keys. The default is 4096.
- `acme_certificate_key_name`: The basename for storing the keys and certificates. The default is the first domain specified, with `*` replaced by `_`.
- `acme_certificate_keys_path`: Where the keys and certificates are stored. Default value is `"keys/"`.
- `acme_certificate_keys_old_path`: Where old keys and certificates should be copied to; used in case `acme_certificate_keys_old_store` is true. Default value is `"keys/old/"`.
- `acme_certificate_keys_old_store`: If set to `true`, will make copies of old keys and certificates. The copies will be stored in the directory specified by `acme_certificate_keys_old_store`. Default value is `false`.
- `acme_certificate_keys_old_prepend_timestamp`: Whether copies of old keys and certificates should be prepended by the current date and time. Default value is `false`.
- `acme_certificate_ocsp_must_staple`: Whether a certificate with the OCSP Must Staple extension is requested. Default value is `false`.
- `acme_certificate_agreement`: The terms of service document the user agrees to. Default value is `https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf`.
- `acme_certificate_acme_directory`: The ACME directory to use. Default is `https://acme-v02.api.letsencrypt.org/directory`, which is the current production ACME v2 endpoint of Let's Encrypt.
- `acme_certificate_acme_version`: The ACME directory's version. Default is 2. Use 1 for ACME v1.
- `acme_certificate_challenge`: The challenge type to use. Should be `http-01` for HTTP challenges (needs access to web server) or `dns-01` for DNS challenges (needs access to DNS provider).
- `acme_certificate_root_certificate`: The root certificate for the ACME directory. Default value is `https://letsencrypt.org/certs/isrgrootx1.pem` for the root certificate of Let's Encrypt.
- `acme_certificate_deactivate_authzs`: Whether `authz`s (authorizations) should be deactivated afterwards. Default value is `true`. Set to `false` to be able to re-use `authz`s.
- `acme_certificate_modify_account`: Whether the ACME account should be created (if it doesn't exist) and the contact data (email address) should be updated. Default value is `true`. Set to `false` if you want to use the `acme_account` module to manage your ACME account (not done by this role).
- `acme_certificate_privatekey_mode`: Which file mode to use for the private key file. Default value is `"0600"`, which means read- and writeable by the owner, but not accessible by anyone else (except possibly `root`).
- `acme_certificate_select_chain`: (Only usable with Ansible 2.10+) Must be in the format described [here](https://docs.ansible.com/ansible/2.10/collections/community/crypto/acme_certificate_module.html#parameter-select_chain). Allows to select the certificate chain to be used; `acme_certificate_root_certificate` must be used in conjunction. This can be used for example with [Let's Encrypt](https://community.letsencrypt.org/t/transition-to-isrgs-root-delayed-until-sep-29/125516) to select which root certificate to use. The following configuration makes sure that the IdenTrust cross-signed intermediate is used, which is more compatible for example for older Android versions than the new IRSG root:
  ```.yaml
  acme_certificate_root_certificate: https://letsencrypt.org/certs/trustid-x3-root.pem.txt
  acme_certificate_select_chain:
    - test_certificates: last
      issuer:
        CN: DST Root CA X3
        O: Digital Signature Trust Co.
  ```
  The following configuration selects the new IRSG X1 root:
  ```.yaml
  acme_certificate_root_certificate: https://letsencrypt.org/certs/isrgrootx1.pem
  acme_certificate_select_chain:
    - test_certificates: last
      issuer:
        CN: ISRG Root X1
        O: Internet Security Research Group
  ```

### HTTP Challenges

For HTTP challenges, the following variables define how the challenges can be put onto the (remote) webserver:

- `acme_certificate_server_location`: Location where `.well-known/acme-challenge/` will be served from. Default is `/var/www/challenges`.
- `acme_certificate_http_become`: Argument for `become:` for the `file` and `copy` tasks. Default value is `false`.
- `acme_certificate_http_challenge_user`: The user the challenge files are owned by. Default value is `root`.
- `acme_certificate_http_challenge_group`: The group the challenge files are owned by. Default value is `http`.
- `acme_certificate_http_challenge_folder_mode`: The mode to use for the challenge folder. Default value is `0750` (octal).
- `acme_certificate_http_challenge_file_mode`: The mode to use for the challenge files. Default value is `0640` (octal).

The following subsection shows how to configure [nginx](https://nginx.org/) for HTTP challenges. Configuring other webservers can be done in a similar way.

#### Nginx configuration

Assume that for one of your TLS/SSL protected domains, you use a HTTP-to-HTTPS redirect. Let's assume it looks like this:

    server {
        listen       example.com:80;
        server_name  example.com *.example.com;
        return 301   https://www.example.com$request_uri;
    }

To allow the `acme_certificate` role to put something at `http://*.example.com/.well-known/acme-challenge/`, you can change this to:

    server {
        listen       example.com:80;
        server_name  example.com *.example.com;
        location /.well-known/acme-challenge/ {
            alias /var/www/challenges/;
            try_files $uri =404;
        }
        location / {
            return 301   https://www.example.com$request_uri;
        }
    }

With this nginx config, all other URLs on `*.example.com` and `example.com` are still redirected, while everything in `*.example.com/.well-known/acme-challenge/` is served from `/var/www/challenges`. When adjusting the location of `/var/www/challenges`, you must also change `acme_certificate_server_location`.

You can even improve on this by redirecting all URLs in `*.example.com/.well-known/acme-challenge/` which do not resolve to a valid file in `/var/www/challenges` to your HTTPS server as well. One way to do this is:

    server {
        listen       example.com:80;
        server_name  example.com *.example.com;
        location /.well-known/acme-challenge/ {
            alias /var/www/challenges/;
            try_files $uri @forward_https;
        }
        location @forward_https {
            return 301   https://www.example.com$request_uri;
        }
        location / {
            return 301   https://www.example.com$request_uri;
        }
    }

With this config, if `/var/www/challenges/` is empty, your HTTP server will behave as if the `/.well-known/acme-challenge/` location isn't specified.

### DNS Challenges

If DNS challenges are used, the following variables define how the challenges can be fulfilled:

- `acme_certificate_dns_provider`: must be one of `route53`, `hosttech`, and `ns1`. Each needs more information:
  - For `route53` (Amazon Route 53), the credentials must be passed as `acme_certificate_aws_access_key` and `acme_certificate_aws_secret_key`.
  - For `hosttech` (hosttech GmbH, requires external [hosttech_dns_record module](https://github.com/felixfontein/ansible-hosttech)).
  - For `ns1` ([ns1.com](https://ns1.com)) the key for your API account must be passed as `acme_certificate_ns1_secret_key`. Also it depends on external module `ns1_record`. Assuming default directory structure and settings, you may need download 2 files into machine where playbook executed:

  ```bash
  curl --create-dirs -L -o ~/.ansible/plugins/module_utils/ns1.py https://github.com/ns1/ns1-ansible-modules/raw/master/module_utils/ns1.py

CHANGELOG.md
# Changelog for acme_certificate

## Version 1.1.2 (2023-07-21)

- Add deprecation marker.

## Version 1.1.1 (2020-05-22)

- Linting, to make Galaxy more happy. (ansible-lint does not like missing modules. This might get better with collections.)

## Version 1.1.0 (2020-05-22)

- Added better namespacing for role parameters; all role parameters now start with `acme_certificate_`. The old, shorter names can still be used for now. Support for them will be dropped in version 2.0.0, to be released later this year.
- Dropped support for GCDNS (which never worked).
- Support for DNS provider NS1 for DNS challenges (thanks to @timelapserduck).
- Lint YAML files (thanks to @pgporada).
- Allow `key_path` to not have trailing slash (thanks to @nwmcsween).
- Fix curve used for P-256.
- Require Ansible 2.8.3.

## Version 1.0 (2019-07-01)

First version published on Ansible Galaxy.


roles/letsencrypt/code/acme_lib.py
"""
Compact ACME library.

Based on the py3 branch of https://github.com/collinanderson/acme-tiny,
which is based on https://github.com/diafygi/acme-tiny.

This code is licensed under the MIT license; see LICENSE for details.

The original acme-tiny code is Copyright (c) 2015 Daniel Roesler
"""

import base64
import binascii
import copy
import hashlib
import json
import os
import re
import subprocess
import sys
import textwrap
import time
try:
    from urllib.request import urlopen, Request
except ImportError:  # Python 2
    from urllib2 import urlopen, Request


staging_ca = "https://acme-staging.api.letsencrypt.org"
default_ca = "https://acme-v01.api.letsencrypt.org"
default_ca_staging = "https://acme-staging.api.letsencrypt.org"
default_intermediate_url = "https://letsencrypt.org/certs/lets-encrypt-x3-cross-signed.pem"
default_root_url = "https://letsencrypt.org/certs/isrgrootx1.pem"
ca_agreement = "https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf"
ca_agreement_redirect_pattern = '{}/terms'

user_agent = "acme-compact"

# #####################################################################################################
# # Helper functions


def _b64(b):
    """Helper function base64 encode for jose spec."""
    return base64.urlsafe_b64encode(b).decode('utf8').replace("=", "")


def _run_openssl(args, input=None):
    """Execute OpenSSL with the given arguments. Feeds input via stdin if given."""
    if input is None:
        proc = subprocess.Popen(["openssl"] + list(args), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        out, err = proc.communicate()
    else:
        proc = subprocess.Popen(["openssl"] + list(args), stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        out, err = proc.communicate(input)
    if proc.returncode != 0:
        raise IOError("OpenSSL Error: {0}".format(err.decode('utf-8')))
    return out


def _get_wellknown_path(domain, token, folder_for_domain):
    """Retrieve path for token file."""
    if callable(folder_for_domain):
        folder = folder_for_domain(domain)
    else:
        folder = folder_for_domain
    return os.path.join(folder, token)


def _request(url, content_type=None):
    headers = {'User-Agent': user_agent}
    if content_type:
        headers['Content-Type'] = content_type
    return Request(url, headers=headers)


# #####################################################################################################
# # Algorithm support


class Algorithm(object):
    """Abstracts an algorithm (RSA or ECC)."""

    def __init__(self, name):
        """Initialize algorithm object."""
        self.name = name

    def __not_implemented(self, method):
        """Helper method to raise not implemented errors."""
        raise Exception("Algorithm {0} does not support {1}!".format(self.name, method))

    def create_key(self, key_length):
        """Create a private key of given length."""
        self.__not_implemented('create_key')


class RSA(Algorithm):
    """Abstracts the RSA algorithm."""

    def __init__(self):
        """Create new RSA algorithm object."""
        super(RSA, self).__init__("RSA")
        self.jws_algorithm = 'RS256'
        self.jws_hash = 'sha256'
        self.jws_hash_bytes = 32

    def create_key(self, key_length):
        """Generate RSA key with given key length."""
        return _run_openssl(['genrsa', str(key_length)]).decode('utf-8')


class ECC(Algorithm):
    """Abstracts Elliptic Curve based algorithms."""

    def __init__(self, curve, openssl_curve, bitlength, jws_algorithm, jws_hash, jws_hash_bytes):
        """Create new ECC algorithm object for given JOSE curve name, OpenSSL curve name, and bit length."""
        super(ECC, self).__init__("ECC-{0}".format(curve))
        self.curve = curve
        self.openssl_curve = openssl_curve
        self.bitlength = bitlength
        self.bytelength = (bitlength + 7) // 8
        self.jws_algorithm = jws_algorithm
        self.jws_hash = jws_hash
        self.jws_hash_bytes = jws_hash_bytes

    def create_key(self, key_length):
        """Generate ECC private key for this curve. The key length is ignored."""
        return _run_openssl(['ecparam', '-name', self.openssl_curve, '-genkey', '-noout']).decode('utf-8')

    def extract_point(self, pub_hex):
        """Extract the public point coordinates from the given hexadecimal description."""
        if len(pub_hex) != 2 * self.bytelength:
            raise ValueError("Key error: public key has incorrect length")
        return pub_hex[:self.bytelength], pub_hex[self.bytelength:]


_ALGORITHMS = {
    'rsa': RSA(),
    'p-256': ECC('p-256', 'prime256v1', 256, 'ES256', 'sha256', 32),
    'p-384': ECC('p-384', 'secp384r1', 384, 'ES384', 'sha384', 48),
    # 'p-521': ECC('p-521', 'secp521r1', 528, 'ES512', 'sha512', 64),
    #          -- P-521 isn't supported yet (on Let's Encrypt staging server);
    #             see https://github.com/letsencrypt/boulder/issues/2217
}


def _get_algorithm(algorithm):
    if algorithm not in _ALGORITHMS:
        raise ValueError("Unknown algorithm '{0}'!".format(algorithm))
    return _ALGORITHMS[algorithm]


# #####################################################################################################
# # Low level functions


def read_stdin():
    """Read data from stdin."""
    if sys.version_info < (3, 0):
        return sys.stdin.read()
    else:
        return sys.stdin.buffer.read()


def write_file(filename, content):
    """Write the contents (string) into the file, encoded with UTF-8."""
    with open(filename, "wb") as f:
        f.write(content.encode('utf-8'))


def create_key(key_length=4096, algorithm="rsa"):
    """Create an RSA key with the given key length in bits."""
    algorithm = _get_algorithm(algorithm)
    return algorithm.create_key(key_length)


def generate_csr(key_filename, config_filename, domains, must_staple=False):
    """Given a private key and a list of domains, create a Certificate Signing Request (CSR).

    ``must_staple```: if set to ``True``, asks for a certificate with OCSP Must Staple enabled.
    """
    # First generate config
    template = """HOME     = .
RANDFILE = $ENV::HOME/.rnd

[req]
distinguished_name = req_DN
req_extensions     = req_SAN

[req_DN]

[req_SAN]
subjectAltName = {0}
"""
    if must_staple:
        # See https://tools.ietf.org/html/rfc7633#section-6 and https://scotthelme.co.uk/ocsp-must-staple/
        template += "1.3.6.1.5.5.7.1.24 = DER:30:03:02:01:05\n"
        # For OpenSSL 1.1.0 or newer, we can use
        #     template += "tlsfeature = status_request\n"
        # instead.
    write_file(config_filename, template.format(','.join(['DNS:{0}'.format(domain) for domain in domains])))
    # Generate CSR
    if key_filename == '/dev/stdin':
        stdin = read_stdin()
        return _run_openssl(['req', '-new', '-sha256', '-key', '/dev/stdin', '-subj', '/', '-config', config_filename], input=stdin).decode('utf-8')
    else:
        return _run_openssl(['req', '-new', '-sha256', '-key', key_filename, '-subj', '/', '-config', config_filename]).decode('utf-8')


def get_csr_as_text(csr_filename):
    """Convert CSR file to plaintext with OpenSSL."""
    return _run_openssl(["req", "-in", csr_filename, "-noout", "-text"]).decode('utf-8')


def parse_account_key(account_key):
    """Parse account RSA private key to get public key.

    Returns five variables (account_key_type, account_key, account_key_algorithm, header, thumbprint)
    needed for other low-level functions.
    """
    sys.stderr.write("Parsing account key...")
    account_key_type = None
    with open(account_key, "r") as f:
        for line in f:
            m = re.match(r"^\s*-{5,}BEGIN\s+(EC|RSA)\s+PRIVATE\s+KEY-{5,}\s*$", line)
            if m is not None:
                account_key_type = m.group(1).lower()
                break
    if account_key_type not in ("rsa", "ec"):
        raise ValueError("Unknown key type '{0}'.".format(account_key_type))
    out = _run_openssl([account_key_type, "-in", account_key, "-noout", "-text"]).decode('utf8')
    if account_key_type == "rsa":
        pub_hex, pub_exp = re.search(r"modulus:\n\s+00:([a-f0-9\:\s]+?)\npublicExponent: ([0-9]+)", out, re.MULTILINE | re.DOTALL).groups()
        pub_mod = binascii.unhexlify(re.sub(r"(\s|:)", "", pub_hex))
        pub_mod64 = _b64(pub_mod)
        pub_exp = "{0:x}".format(int(pub_exp))
        pub_exp = "0{0}".format(pub_exp) if len(pub_exp) % 2 else pub_exp
        pub_exp64 = _b64(binascii.unhexlify(pub_exp))
        algorithm = _get_algorithm('rsa')
        header = {
            "alg": algorithm.jws_algorithm,
            "jwk": {
                "kty": "RSA",
                "e": pub_exp64,
                "n": pub_mod64,
            },
        }
    else:
        pub_data = re.search(
            r"pub:\s*\n\s+04:([a-f0-9\:\s]+?)\nASN1 OID: (\S+)\nNIST CURVE: (\S+)", out, re.MULTILINE | re.DOTALL)
        if pub_data is None:
            raise ValueError("Invalid or incompatible ECC key.")
        pub_hex = binascii.unhexlify(re.sub(r"(\s|:)", "", pub_data.group(1)))
        curve = pub_data.group(3).lower()
        algorithm = _get_algorithm(curve)
        x, y = algorithm.extract_point(pub_hex)
        header = {
            "alg": algorithm.jws_algorithm,
            "jwk": {
                "kty": "EC",
                "crv": curve.upper(),
                "x": _b64(x),
                "y": _b64(y),
            },
        }
    accountkey_json = json.dumps(header['jwk'], sort_keys=True, separators=(',', ':'))
    thumbprint = _b64(hashlib.sha256(accountkey_json.encode('utf8')).digest())
    sys.stderr.write(" ok ")
    sys.stderr.flush()
    return account_key_type, account_key, algorithm, header, thumbprint


def _lookup_directory(CA, *keys):
    """Helper function to look up the correct URL(s) in the server's directory.

    Returns a replay nonce (or None) and the URLs associated to the keys:
    ``(nonce, url_1, url_2, ...)``
    """
    result = urlopen(_request(CA + "/directory"))
    data = json.loads(result.read().decode('utf8'))
    urls = [data.get(key, CA + '/acme/' + key) for key in keys]
    nonce = result.headers['Replay-Nonce']
    return tuple([nonce] + urls)


def _send_signed_request(payload, header, CA, account_key_type, account_key, account_key_algorithm, key=None, url=None):
    """Helper function make signed requests. Either ``key`` or ``url`` must be specified."""
    # Make sure we know the URL, and figure out nonce_url (and see if we get a nonce as well)
    assert key is not None or url is not None
    if url is None:
        nonce, url, nonce_url = _lookup_directory(CA, key, 'new-nonce')
    else:
        nonce, nonce_url = _lookup_directory(CA, 'new-nonce')
    # If we didn't already got a nonce, ask nonce_url
    if nonce is None:
        nonce = urlopen(_request(nonce_url)).headers['Replay-Nonce']
    # Prepare JOSE object
    payload64 = _b64(json.dumps(payload).encode('utf8'))
    protected = copy.deepcopy(header)
    protected.update({"nonce": nonce})
    protected64 = _b64(json.dumps(protected).encode('utf8'))
    out = _run_openssl(["dgst", "-{0}".format(account_key_algorithm.jws_hash), "-sign", account_key], "{0}.{1}".format(protected64, payload64).encode('utf8'))
    if account_key_type == 'ec':
        out = _run_openssl(["asn1parse", "-inform", "DER"], input=out).decode("utf8")
        sig = re.findall(r"prim:\s+INTEGER\s+:([0-9A-F]{%s})\n" % (2 * account_key_algorithm.jws_hash_bytes), out)
        if len(sig) != 2:
            raise Exception("Failed to generate signature; cannot parse DER output:\n\n{0}".format(out))
        out = binascii.unhexlify(sig[0]) + binascii.unhexlify(sig[1])
    data = json.dumps({
        "header": header,
        "protected": protected64,
        "payload": payload64,
        "signature": _b64(out),
    })
    # Send JOSE object
    try:
        resp = urlopen(_request(url, 'application/jose+json'), data.encode('utf8'))
        return resp.getcode(), resp.read()
    except IOError as e:
        return getattr(e, "code", None), getattr(e, "read", e.__str__)()


def parse_csr(csr):
    """Parse a Certificate Signing Request (CSR).

    Returns the list of domains this CSR affects.
    """
    out = get_csr_as_text(csr)
    domains = set([])
    common_name = re.search(r"Subject:.*? CN\s*=\s*([^\s,;/]+)", out)
    if common_name is not None:
        domains.add(common_name.group(1))
    for subject_alt_names in re.finditer(r"X509v3 Subject Alternative Name: (?:critical)?\n +([^\n]+)\n", out, re.MULTILINE | re.DOTALL):
        for san in subject_alt_names.group(1).split(", "):
            if san.startswith("DNS:"):
                domains.add(san[4:])
    return sorted(domains)


def register_account(header, CA, account_key_type, account_key, account_key_algorithm, email_address=None, telephone=None):
    """Create account on CA server.

    Return True if the account was created and False if it already exists.
    Raises an exception in case of errors.
    """
    argreement = ca_agreement
    try:
        resp = urlopen(_request(ca_agreement_redirect_pattern.format(CA)))
        argreement = resp.url
    except IOError as e:
        sys.stderr.write("Retrieving agreement failed: {0}\n".format(e.message))
    data = {
        "resource": "new-reg",
        "agreement": argreement,
    }
    contacts = []
    if email_address is not None:
        contacts.append("mailto:{0}".format(email_address))
    if telephone is not None:
        contacts.append("tel:{0}".format(telephone))
    if len(contacts) > 0:
        data["contact"] = contacts
    code, result = _send_signed_request(data, header, CA, account_key_type, account_key, account_key_algorithm, key="new-reg")
    if code == 201:
        return True
    elif code == 409:
        return False
    else:
        raise ValueError("Error registering: {0} {1}".format(code, result))


def get_challenge(domain, header, CA, account_key_type, account_key, account_key_algorithm, thumbprint):
    """Retrieve challenge for a domain.

    Returns the challenge object, the challenge token as well as the
    content for the token file.
    """
    # get new challenge
    code, result = _send_signed_request({
        "resource": "new-authz",
        "identifier": {"type": "dns", "value": domain},
    }, header, CA, account_key_type, account_key, account_key_algorithm, key="new-authz")
    if code != 201:
        raise ValueError("Error registering: {0} {1}".format(code, result))

    # make the challenge file
    challenge = [c for c in json.loads(result.decode('utf8'))['challenges'] if c['type'] == "http-01"][0]
    challenge['token'] = re.sub(r"[^A-Za-z0-9_\-]", "_", challenge['token'])
    keyauthorization = "{0}.{1}".format(challenge['token'], thumbprint)
    return challenge, challenge['token'], keyauthorization


def get_wellknown_url(domain, token):
    """Return the URL for the token file on the server."""
    return "http://{0}/.well-known/acme-challenge/{1}".format(domain, token)


def check_challenge(domain, token, keyauthorization):
    """Check whether the token is correctly placed on the server.

    Returns True in case it is, and False in case it is not.
    """
    # check that the file is in place
    wellknown_url = get_wellknown_url(domain, token)
    try:
        resp = urlopen(_request(wellknown_url))
        return resp.read().decode('utf8').strip() == keyauthorization
    except IOError:
        return False


def notify_challenge(domain, header, CA, account_key_type, account_key, account_key_algorithm, challenge, keyauthorization):
    """Notify the CA server that the token files are available on the webserver."""
    # notify challenge are met
    code, result = _send_signed_request({
        "resource": "challenge",
        "keyAuthorization": keyauthorization,
    }, header, CA, account_key_type, account_key, account_key_algorithm, url=challenge['uri'])
    if code != 202:
        raise ValueError("Error triggering challenge: {0} {1}".format(code, result))


def check_challenge_verified(domain, challenge, wait=True):
    """Check whether the challenge has been verified by the CA server.

    Returns True in case it was successfully verified, and False in case
    the verification is not done yet. Raises an exception in case the
    verification failed.

    If wait is set to True (default), the function will not return False
    but instead loop until it either returns True or it raises an exception.
    """
    while True:
        try:
            resp = urlopen(_request(challenge['uri']))
            challenge_status = json.loads(resp.read().decode('utf8'))
        except IOError as e:
            raise ValueError("Error checking challenge: {0} {1}".format(e.code, json.loads(e.read().decode('utf8'))))
        if challenge_status['status'] == "pending":
            if wait:
                time.sleep(2)
            else:
                return False
        elif challenge_status['status'] == "valid":
            return True
        else:
            raise ValueError("{0} challenge did not pass: {1}".format(domain, challenge_status))


def retrieve_certificate(csr, header, CA, account_key_type, account_key, account_key_algorithm):
    """Retrieve the certificate from the CA server."""
    sys.stderr.write("Signing certificate...")
    csr_der = _run_openssl(["req", "-in", csr, "-outform", "DER"])
    code, result = _send_signed_request({
        "resource": "new-cert",
        "csr": _b64(csr_der),
    }, header, CA, account_key_type, account_key, account_key_algorithm, key="new-cert")
    if code != 201:
        raise ValueError("Error signing certificate: {0} {1}".format(code, result))
    return """-----BEGIN CERTIFICATE-----\n{0}\n-----END CERTIFICATE-----\n""".format("\n".join(textwrap.wrap(base64.b64encode(result).decode('utf8'), 64)))


def download_certificate(url):
    """Download a certificate (as a file) from the CA server."""
    try:
        resp = urlopen(_request(url))
        if resp.getcode() != 200:
            raise ValueError("Cannot retrieve certificate (status code {0}; message: {1})".format(resp.getcode(), resp.read()))
        return resp.read().decode('utf-8').strip()
    except IOError as e:
        raise ValueError("Cannot retrieve certificate ({0})".format(str(e)))


# #####################################################################################################
# # High level functions


def serialize_state(state):
    """Serialize the state as a string."""
    return json.dumps(state, sort_keys=True)


def deserialize_state(serialized_state):
    """Deserialize the given serialized state.

    Raises exception in case this is not a valid state.
    """
    result = json.loads(serialized_state)
    if type(result) != dict or 'account_key' not in result or 'account_key_type' not in result or 'account_key_algorithm' not in result or 'header' not in result or 'thumbprint' not in result or 'CA' not in result or 'challenges' not in result:
        raise ValueError("Not a valid serialized state!")
    return result


def get_challenges(account_key, csr, CA, email_address=None, telephone=None):
    """Set up the account and retrieve challenges from CA server.

    Returns a state object.
    """
    account_key_type, account_key, account_key_algorithm, header, thumbprint = parse_account_key(account_key)
    # find domains
    domains = parse_csr(csr)
    # get the certificate domains and expiration
    register_account(header, CA, account_key_type, account_key, account_key_algorithm, email_address=email_address, telephone=telephone)
    challenges = []
    # verify each domain
    for domain in domains:
        challenge, token, keyauthorization = get_challenge(domain, header, CA, account_key_type, account_key, account_key_algorithm, thumbprint)
        challenges.append({'domain': domain, 'challenge': challenge, 'token': token, 'keyauthorization': keyauthorization})
    return {'account_key_type': account_key_type, 'account_key_algorithm': account_key_algorithm, 'account_key': account_key, 'header': header, 'thumbprint': thumbprint, 'CA': CA, 'challenges': challenges}


def write_challenges(state, folder_for_domain):
    """Write challenge files to disk.

    If the folder_for_domain parameter is a callable, it is expected to
    return a path when called with a single parameter, which will be the
    domain name. Otherwise, it is assumed to be a string.
    """
    challenges = state['challenges']
    for challenge_entry in challenges:
        domain = challenge_entry['domain']
        token = challenge_entry['token']
        keyauthorization = challenge_entry['keyauthorization']
        wellknown_path = _get_wellknown_path(domain, token, folder_for_domain)
        write_file(wellknown_path, keyauthorization)


def remove_challenges(state, folder_for_domain):
    """Remove the challenge files from disk.

    See documentation of write_challenges() for explanation
    of folder_for_domain.
    """
    challenges = state['challenges']
    for challenge_entry in challenges:
        domain = challenge_entry['domain']
        token = challenge_entry['token']
        wellknown_path = _get_wellknown_path(domain, token, folder_for_domain)
        os.remove(wellknown_path)


def verify_challenges(state):
    """Verify that the challenge files are available on the web server with HTTP."""
    challenges = state['challenges']
    for challenge_entry in challenges:
        domain = challenge_entry['domain']
        token = challenge_entry['token']
        keyauthorization = challenge_entry['keyauthorization']
        if not check_challenge(domain, token, keyauthorization):
            raise ValueError("Couldn't download challenge file at {0}".format(get_wellknown_url(domain, token)))


def notify_challenges(state):
    """Notify the CA server that the challenges are ready."""
    challenges = state['challenges']
    for challenge_entry in challenges:
        domain = challenge_entry['domain']
        keyauthorization = challenge_entry['keyauthorization']
        challenge = challenge_entry['challenge']
        notify_challenge(domain, state['header'], state['CA'], state['account_key_type'], state['account_key'], state['account_key_algorithm'], challenge, keyauthorization)


def check_challenges(state, csr, inform=None):
    """Check the CA server for challenge results, and retrieves the certificate.

    In case inform is specified, it is called with the domain name as the only argument
    for every successfully verified domain.

    When all domains are verified, the certificate is obtained from the CA server and
    returned as a string.
    """
    challenges = state['challenges']
    for challenge_entry in challenges:
        domain = challenge_entry['domain']
        challenge = challenge_entry['challenge']
        check_challenge_verified(domain, challenge, wait=True)
        if callable(inform):
            inform(domain)
    return retrieve_certificate(csr, state['header'], state['CA'], state['account_key_type'], state['account_key'], state['account_key_algorithm'])
